pallets/liquidity-pools-gateway/src/message_processing.rs

```diff
@@ -0,0 +1,539 @@
+	}
+
+	/// Upserts an inbound entry for a particular message, increasing the
+	/// relevant counts accordingly.
+	pub(crate) fn upsert_pending_entry(
+		message_proof: Proof,
+		router_id: T::RouterId,
+		new_inbound_entry: InboundEntry<T>,
+	) -> DispatchResult {
+		PendingInboundEntries::<T>::try_mutate(message_proof, router_id, |storage_entry| {
+			match storage_entry {
+				None => {
+					*storage_entry = Some(new_inbound_entry);
+
+					Ok::<(), DispatchError>(())
+				}
+				Some(stored_inbound_entry) => {
+					stored_inbound_entry.pre_dispatch_update(new_inbound_entry)
+				}
+			}
+		})
+	}
+
+	/// Creates, validates and upserts the inbound entry.
+	fn validate_and_upsert_pending_entries(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message: T::Message,
+		message_proof: Proof,
+		router_id: T::RouterId,
+	) -> DispatchResult {
+		let inbound_entry: InboundEntry<T> = (inbound_processing_info, message).into();
+
+		inbound_entry.validate(&inbound_processing_info, &router_id)?;
+
+		Self::upsert_pending_entry(message_proof, router_id, inbound_entry)?;
+
+		Ok(())
+	}
+
+	/// Checks if the number of proofs required for executing one message
+	/// were received, and if so, decreases the counts accordingly and executes
+	/// the message.
+	pub(crate) fn execute_if_requirements_are_met(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message_proof: Proof,
+	) -> DispatchResult {
+		let mut message = None;
+		let mut votes = 0;
+
+		for router_id in &inbound_processing_info.router_ids {
+			match PendingInboundEntries::<T>::get(message_proof, router_id) {
+				// We expected one InboundEntry for each router, if that's not the case,
+				// we can return.
+				None => return Ok(()),
+				Some(stored_inbound_entry) => match stored_inbound_entry {
+					InboundEntry::Message(message_entry) => message = Some(message_entry.message),
+					InboundEntry::Proof(proof_entry) => {
+						if proof_entry
+							.has_valid_vote_for_session(inbound_processing_info.current_session_id)
+						{
+							votes.ensure_add_assign(1)?;
+						}
+					}
+				},
+			};
+		}
+
+		if votes < inbound_processing_info.expected_proof_count_per_message {
+			return Ok(());
+		}
+
+		match message {
+			Some(msg) => {
+				Self::execute_post_voting_dispatch(&inbound_processing_info, message_proof)?;
+
+				T::InboundMessageHandler::handle(
+					inbound_processing_info.domain_address.clone(),
+					msg,
+				)
+			}
+			None => Ok(()),
+		}
+	}
+
+	/// Decreases the counts for inbound entries and removes them if the
+	/// counts reach 0.
+	pub(crate) fn execute_post_voting_dispatch(
+		inbound_processing_info: &InboundProcessingInfo<T>,
+		message_proof: Proof,
+	) -> DispatchResult {
+		for router_id in &inbound_processing_info.router_ids {
+			PendingInboundEntries::<T>::try_mutate(message_proof, router_id, |storage_entry| {
+				match storage_entry {
+					None => {
+						Err::<(), DispatchError>(Error::<T>::PendingInboundEntryNotFound.into())
```

Comment:
True, this can only happen if one is to call it right after an initial message is received is something, but also for consistency's sake, I would like to have a clear error for this case.

Reply:
As you think is better 👍🏻, could you comment saying that this can not be produced?

Reply:
Sure, thank you.